File: python/RemoveNthNodeFromEndofList.py


class Solution:
    # @return a ListNode
    def removeNthFromEnd(self, head, n):
        tail = head
        while n :
            tail, n = tail.next, n-1
        prev, cur = head, head
        while tail :
            tail, cur, prev = tail.next, cur.next, cur
        if prev == cur :
            return prev.next
        else :
            prev.next = cur.next
            return head
    def printList(self, head) :
        while head :
            print head.val
            head = head.next


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

L1 = ListNode(1)
L2 = ListNode(2)
L3 = ListNode(3)
L4 = ListNode(4)
L5 = ListNode(5)
L1.next = L2
L2.next = L3
L3.next = L4
L4.next = L5

sol = Solution()
k = sol.removeNthFromEnd(L1, 2)
sol.printList(k)

File: cpluplus/Remove_Nth_Node_From_End_of_List.cc
/*
 * author: yuandx
 * create: 2015-11-17
 * email: yuandx@mvad.com
 */

#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <climits>
#include <string>
#include <utility>
#include <functional>
#include <algorithm>
#include <stdio.h>
#include <sstream>
#include <unordered_map>
#include <unordered_set>

using std::unordered_map;
using std::unordered_set;
using std::ostringstream;
using std::istringstream;
using std::stringstream;
using std::cout;
using std::cin;
using std::endl;
using std::vector;
using std::map;
using std::set;
using std::pair;
using std::sort;
using std::string;
using std::for_each;

//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
public:
  ListNode* removeNthFromEnd(ListNode* head, int n)
  {
    ListNode *ln = NULL, *tail = head, *cur = head;
    for (int i = 0; i < n; i++)
    {
      tail = tail->next;
    }
    for (; tail != NULL; tail = tail->next)
    {
      ln = cur;
      cur = cur->next;
    }
    if (ln == NULL)
    {
      head = head->next;
    }
    else
    {
      ln->next = cur->next;
    }
    cur->next = NULL;
    return head;
  }
};

int main()
{

  return 0;
}
